core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -487,6 +487,33 @@ class TransactionCoordinator(brokerId: Int,
               info(s"Aborting sending of transaction markers and returning $error error to client for $transactionalId's EndTransaction request of $txnMarkerResult, " +
                 s"since appending $newMetadata to transaction log with coordinator epoch $coordinatorEpoch failed")
 
+              txnManager.getTransactionState(transactionalId).right.foreach {
```

Comment:
Took me a while to remember this issue... So basically the coordinator has decided to abort a transaction and has bumped the epoch. However, when it tries to write the updated state to the log, it fails, which leaves us in an inconsistent state. Of course failing to write to the log doesn't necessarily mean that the entry wasn't appended. In fact, it could still become committed. There is no way to take the write back once it gets to the log. Hence I'm a little hesitant about the logic to revert to the previous epoch in this case. Would it still be possible for the fenced producer to make progress with the old epoch after reverting? Perhaps another idea would be to keep the epoch bumped in memory, but remember the fact that the write had failed. So the next time we go to retry the abort, we do not need to bump the epoch again. Does that make sense?

Reply:
That's a good point.. the append returning "fail" does not mean that the append did not go through. I think the alternative idea would work better: basically, whenever the coordinator decides to abort a txn, it can mark in memory the current epoch as aborting and whenever it (re-)tries to write the prepare-abort entry it always just use the epoch+1 until the write goes through and we can reset that aborting marker.

Reply:
Thanks for the suggestion, @hachikuji! I wound up taking that approach.